### PR TITLE
Add developer guide

### PR DESCRIPTION
**Pull request by milesfrain**

### CODE REVIEW

File: INSTALL.md
Comment (milesfrain) [inline INSTALL.md:38]: Is the `purescript:RELEASE` flag still necessary? I removed those from the new instructions since omitting it wasn't causing issues for me.

File: DEVELOPER_GUIDE.md
Comment (hdgarrood) [inline DEVELOPER_GUIDE.md:101]: I don't really want to maintain instructions for particular IDE projects or editors here. There's nothing particularly weird about the PureScript compiler that should necessitate tooling or editor-specific guides, and if we're including instructions for HIE and VSCode, why should we not also include instructions for emacs, vim, ghcid, ghcide, etc.? I think this file will very quickly get out of hand (and probably also out of date) with this approach. This sort of documentation is best placed in the particular tooling/editor plugin projects themselves, I think.

File: DEVELOPER_GUIDE.md
Comment (hdgarrood) [inline DEVELOPER_GUIDE.md:11]: This command doesn't update stack, it only updates stack's package index.

File: DEVELOPER_GUIDE.md
Comment (hdgarrood) [inline DEVELOPER_GUIDE.md:18]: If people are trying to contribute, they should almost always be starting from `master`, so I think I'd prefer to remove this bit.

File: INSTALL.md
Comment (hdgarrood) [reply INSTALL.md:38]: Can you please leave this file as-is? This file is really intended for people wanting to install the compiler, not to contribute to it. The RELEASE flag is needed for ensuring that the output of `purs --version` is correct.

File: DEVELOPER_GUIDE.md
Comment (hdgarrood) [inline DEVELOPER_GUIDE.md:1]: I'm not sure this should be a new file; isn't it targeted at the same audience that the existing CONTRIBUTING.md file is?

Comment (milesfrain) [reply DEVELOPER_GUIDE.md:1]: The CONTRIBUTING.md file seemed for focused on policies rather than getting things running (apart from the "tests" section), but I'm happy to move this content there.

File: INSTALL.md
Comment (milesfrain) [reply INSTALL.md:38]: For users who just want to install the compiler (without contributing), would it be simpler to replace that section with:
```
stack install purescript
```
I'm having trouble envisioning what target audience would prefer the above instructions, since both of these approaches rely on trusting stack + hackage.

Comment (hdgarrood) [reply INSTALL.md:38]: No, `stack install purescript` would behave differently to what we have there; it tries to install it in a particular snapshot and doesn't use `purescript`'s own stack.yaml file.

File: DEVELOPER_GUIDE.md
Comment (milesfrain) [reply DEVELOPER_GUIDE.md:71]: There was some confusion about this behavior in one of the meetings.
So I think it would be helpful to note that these faster rebuilds are available in stack 2.x.

Comment (hdgarrood) [reply DEVELOPER_GUIDE.md:71]: Can you be more specific? I don't see how it matters; if you need to do a profiling build, then there's not really any way around that; you just have to wait for as long as it takes.

File: DEVELOPER_GUIDE.md
Comment (milesfrain) [reply DEVELOPER_GUIDE.md:101]: I'll move this IDE section to a Discourse post.

File: DEVELOPER_GUIDE.md
Comment (milesfrain) [reply DEVELOPER_GUIDE.md:18]: Motivations for this section are:
* Letting users profile or fix an issue at a specific release. For example they may encounter new bugs if `master` contains breaking changes.
* Describe how to install a particular release (absorbing the removed INSTALL.md content)

But maybe I'm underestimating the abilities of readers to navigate releases.

Comment (hdgarrood) [reply DEVELOPER_GUIDE.md:18]: There's not much point in profiling or fixing an issue at a specific release, though, because it won't be merged if it can't be applied to master, or if the problem no longer exists in master.

File: DEVELOPER_GUIDE.md
Comment (milesfrain) [reply DEVELOPER_GUIDE.md]: The issue was that build workspaces were shared, so toggling build modes acted like clean rebuilds.
1. Standard build - slow
1. Standard build - fast
1. Profiling build - slow
1. Profiling build - fast
1. Standard build - slow (previous work clobbered by recent profiling build)
1. Profiling build - slow (previous work clobbered by recent standard build)

**Comment by milesfrain** [comment]
It looks like the only merge conflict are these lines at the end of the file, which aren't even displayed markdown:
```
[the #purescript channel in FP Slack]: https://functionalprogramming.slack.com/
[Discourse]: https://discourse.purescript.org/
```
See https://github.com/purescript/purescript/pull/3924/files#r515518554

**Comment by hdgarrood** [comment]
The main thing stopping me from approving this right now is that I think the content overlaps with too much of what's already in this file. I can help with that after 0.14.0 is released (since that's what I want to focus my effort on for now).

**Comment by JordanMartinez** [comment]
One other note worth adding to this is [haskell-language-server](https://github.com/haskell/haskell-language-server#installation)

**Comment by milesfrain** [comment]
> I can help with that after 0.14.0 is released (since that's what I want to focus my effort on for now).

@hdgarrood, Is now a good time to work on updating these instructions? I'm happy to collaborate on that.

**Comment by hdgarrood** [comment]
Sorry, I don’t have the bandwidth for this right now. I don’t want to stand in the way of merging this so I will leave it up to other compiler maintainers to make the judgement of when the duplication has been addressed well enough that this can be merged.

File: CONTRIBUTING.md
Comment (JordanMartinez) [inline CONTRIBUTING.md:62]: As of v0.14.2, this is no longer relevant. Can we drop this? Anyone contributing to the repo won't need to be concerned about this anymore.

File: CONTRIBUTING.md
Comment (JordanMartinez) [inline CONTRIBUTING.md:88]: When I tried this on my version of stack, v2.5.1, it wouldn't do anything. I had to pass either a script or a heredoc to make it run something.

I also think it weird that this appears under the 'install' header despite being described as 'testing' the build locally.

File: CONTRIBUTING.md
Comment (JordanMartinez) [inline CONTRIBUTING.md:91]: Should we drop this entirely and instead refer people to docs on that command?

File: CONTRIBUTING.md
Comment (JordanMartinez) [inline CONTRIBUTING.md:95]: I'm always hesitant to use this command because of the global installation. Shouldn't this be covered in `INSTALL.md` instead?

File: CONTRIBUTING.md
Comment (JordanMartinez) [inline CONTRIBUTING.md:140]:
```diff
- stack install profiteur
+ stack build --copy-compiler-tool profiteur
```

File: CONTRIBUTING.md
Comment (JordanMartinez) [inline CONTRIBUTING.md:131]:
```diff
- stack install ghc-prof-flamegraph
+ stack build --copy-compiler-tool ghc-prof-flamegraph
```

**Review by JordanMartinez — CHANGES_REQUESTED** [review]
I think the profiling guide is worth keeping in this PR. I think more installation things should be left out and that we should instead document that in the `INSTALL.md` file. I also think that we should explain less of `stack` and just direct users to its documentation. Having a short "here's the commands you would run in bash" seems fine to me, but not necessarily explaining what they do and why as they can read that in the stack docs.

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md:131]: `stack install` copies to:
```
~/.local/bin/
```

`stack build --copy-compiler-tool` copies to:
```
/.stack/compiler-tools/x86_64-linux-tinfo6/ghc-8.10.4/bin/
```

Is there a way to point to the `compiler-tools` location when running the following command in this readme? Is it necessary to have per-compiler versions of this tool?

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md:88]: `stack exec bash` should create a sub-shell with the recently-built purs version.

```
➜  complier git:(master) ✗ purs --version
0.14.2

➜  complier git:(master) ✗ stack exec bash

miles@x1c5:~/projects/purescript/complier$ purs --version
0.14.2 [development build; commit: f1953214775945b65ba53ae903b4238c352dcd29 DIRTY]

miles@x1c5:~/projects/purescript/complier$ exit

➜  complier git:(master) ✗ purs --version 
0.14.2
```

Maybe this section could be named something other than 'Install', such as 'Integration Testing'. Perhaps change the original 'Testing` header to 'Unit Testing'.

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md:91]: I don't think this usage info is described effectively in the [stack docs](https://docs.haskellstack.org/en/stable/GUIDE/#stack-exec).

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md:95]: I figure some contributors are editing the compiler to fix an issue that's bothering them, and they may want to continue using using their patched version in the meantime while waiting for the next release (without having to run `stack exec` in each new shell). At least that was my use case when writing this guide.

My interpretation of `INSTALL.md` is that it's for folks who just want to grab the latest official version. It seems more efficient to let these two lines live here, rather than edit the [related section](https://github.com/purescript/purescript/blob/master/INSTALL.md#compiling-from-source) of that guide to also include additional instructions for installing untagged versions.

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:131]: > Is it necessary to have per-compiler versions of this tool?

I'm not sure, but it would remove the surprise that people may have when they find that the tool no longer works when using it with other Haskell projects they have on their local computer. I think the main reason to not use `stack install` is that there isn't a corresponding `stack uninstall`. 

If this change was made, the rest of the command would need to be updated to use `stack exec bash`:
```
stack build --copy-compiler-tool ghc-prof-flamegraph
stack exec bash <<EOF
ghc-prof-flamegraph purs.prof
EOF
firefox purs.svg
```

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:88]: Huh... Seems there's something going on with my terminal. I was able to reproduce most of the above behavior, but the info before the `$` part didn't change after running `stack exec bash`

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:91]: Hm... It does mention some part of the idea [here](https://docs.haskellstack.org/en/stable/GUIDE/#exec), but it's still not as clear as what you've written. 

I'd rather fix this upstream. Would you be up for submitting a PR to that repo to update the docs to clarify that? Then we don't have to explain it here when it should really be explained there.

> Remember that `stack exec` must be launched from the same directory that you ran `stack build` from.

I don't think that's quite true. I think it works as long as you are in a subdirectory of the root folder that contains the stack project.

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:95]: Good point. Maybe add a note here that one should use this 'global install' only temporarily for the purpose you mentioned above and that they can remove it by running `stack path --local-bin` to see which folder stores the binary and delete the binary once done?

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:88]: I think the header should be `Exploring and testing a locally compiled version of PureScript` because the purpose of this section is to allow others to just try things and see what happens.

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:91]: Or perhaps we just say

> f you're curious how `stack exec bash` works, run `echo $PATH` then `stack exec bash` then `echo $PATH` again, then `exit`, and finally `echo $PATH` a final time.)

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md]: Changed heading to `Running a locally-compiled version of PureScript`

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md]: > Would you be up for submitting a PR to that repo to update the docs to clarify that?

I made a minor PR to the `stack` docs in https://github.com/commercialhaskell/stack/pull/5574

I'll leave out these two additional `stack exec bash` tips from this guide, and trust that users can figure this out on their own through trial and error:
- You must call `stack exec` from somewhere within the project directory
- You only need to call `stack exec` once

I added some additional tips (e.g. unix `which`) on how to determine which `purs` version is actually being run`.

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md:103]: I think `which purs` beats `stack path --local-bin` for tracking down binaries. Added a note about that utility to this and the previous section.

File: CONTRIBUTING.md
Comment (milesfrain) [reply CONTRIBUTING.md]: This combo works:
```
stack build --copy-compiler-tool profiteur
stack exec -- profiteur purs.prof
```

Updated instructions.

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md]: Thanks. Yeah, using `which` is a better approach than `stack path --local-bin`.

File: CONTRIBUTING.md
Comment (JordanMartinez) [reply CONTRIBUTING.md:103]: Good idea!

**Comment by JordanMartinez** [comment]
In the resources section at the bottom, could you also add a reference to [Haskell Language Server](https://github.com/haskell/haskell-language-server#installation)? That seems to be the direction things are going now.

**Comment by JordanMartinez** [comment]
Otherwise, this looks good to me.

**Review by JordanMartinez — APPROVED** [review]
Thanks! LGTM.

**Comment by JordanMartinez** [comment]
@milesfrain Could you merge this PR with `master`, and then add a changelog entry file in the `CHANGELOG.d` folder? I believe we'll be able to merge this afterwards.

**Comment by JordanMartinez** [comment]
Thanks Miles! @thomashoneyman or @f-f, can one of you review this? If it's good to go, I'd like to merge it.

**Review by f-f — APPROVED** [review]
This looks great! :clap:

**Comment by JordanMartinez** [comment]
Thanks @f-f! And thanks again @milesfrain for working on this!

**Comment by f-f** [comment]
You're welcome :slightly_smiling_face:

**Pull request merged**
